### PR TITLE
Fort15

### PR DESCRIPTION
**Pull request by ghost**
~~not ready to merge yet~~

added
- `RSTIMINC` to `fort.15`
- `NRS` extraction from `NWS` parameter
- unittest to compare simple generated configuration against reference

### CODE REVIEW

**Comment by jreniel** [comment]
I was thinking that instead of :class:`adcircpy.driver.AdcircRun` inheriting from Fort15, we should use composition instead, along with descriptors. It will make it more concise and less confusing, I believe.

**Comment by ghost** [comment]
> I was thinking that instead of :class:`adcircpy.driver.AdcircRun` inheriting from Fort15, we should use composition instead, along with descriptors. It will make it more concise and less confusing, I believe.

That's a good idea, then perhaps we could reuse the class for another configuration file?

**Comment by ghost** [comment]
reopened to keep history with same branch name

**Comment by ghost** [comment]
error with running tests on github workflow: 
https://github.com/JaimeCalzadaNOAA/adcircpy/runs/1365381862
```
======================================================================
ERROR: test_configuration (tests.test_configuration.TestAdcircRun)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/adcircpy/adcircpy/tests/test_configuration.py", line 109, in test_configuration
    driver.write(output_directory, overwrite=True)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/driver.py", line 391, in write
    super().write(
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/fort15.py", line 329, in write
    f.write(self.fort15(runtype))
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/fort15.py", line 110, in fort15
    amp, phase = self.tidal_forcing.tpxo(constituent, vertices)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tides.py", line 628, in tpxo
    return TPXO()
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 31, in __init__
    self._fetch_tpxo_file(prefix, file)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 174, in _fetch_tpxo_file
    a = query_yes_no(q)
  File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 151, in query_yes_no
    choice = input().lower()
EOFError: EOF when reading a line
-------------------- >> begin captured stdout << ---------------------
******* PLEASE READ *******
A function that is being invoked requires the TPXO file.
This software can automatically fetch the TPXO file for you using your internet connection.
You may also cancel this operation and provide the path to the h_tpxo9.v1.nc file using the TPXO_NCFILE environment variable.
By downloading this file and using this software, you are accepting the licensing agreement for the TPXO file found here:
https://drive.google.com/file/d/1f00WojHqu7_VE5Hg9OdiVBjymH76dFCy/view
If you accept the agreement, you may also download the TPXO file from: https://www.dropbox.com/s/uc44cbo5s2x4n93/h_tpxo9.v1.tar.gz?dl=1
Would you like this software to fetch and stage the TPXO file from the internet now?
 [Y/n] 
--------------------- >> end captured stdout << ----------------------

----------------------------------------------------------------------
Ran 63 tests in 4.025s

FAILED (errors=1)
Error: Process completed with exit code 1.
```

is there a way to auto-accept the download?

**Comment by jreniel** [comment]
> error with running tests on github workflow:
> https://github.com/JaimeCalzadaNOAA/adcircpy/runs/1365381862
> 
> ```
> ======================================================================
> ERROR: test_configuration (tests.test_configuration.TestAdcircRun)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/runner/work/adcircpy/adcircpy/tests/test_configuration.py", line 109, in test_configuration
>     driver.write(output_directory, overwrite=True)
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/driver.py", line 391, in write
>     super().write(
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/fort15.py", line 329, in write
>     f.write(self.fort15(runtype))
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/fort15.py", line 110, in fort15
>     amp, phase = self.tidal_forcing.tpxo(constituent, vertices)
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tides.py", line 628, in tpxo
>     return TPXO()
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 31, in __init__
>     self._fetch_tpxo_file(prefix, file)
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 174, in _fetch_tpxo_file
>     a = query_yes_no(q)
>   File "/home/runner/work/adcircpy/adcircpy/adcircpy/forcing/tides/tpxo.py", line 151, in query_yes_no
>     choice = input().lower()
> EOFError: EOF when reading a line
> -------------------- >> begin captured stdout << ---------------------
> ******* PLEASE READ *******
> A function that is being invoked requires the TPXO file.
> This software can automatically fetch the TPXO file for you using your internet connection.
> You may also cancel this operation and provide the path to the h_tpxo9.v1.nc file using the TPXO_NCFILE environment variable.
> By downloading this file and using this software, you are accepting the licensing agreement for the TPXO file found here:
> https://drive.google.com/file/d/1f00WojHqu7_VE5Hg9OdiVBjymH76dFCy/view
> If you accept the agreement, you may also download the TPXO file from: https://www.dropbox.com/s/uc44cbo5s2x4n93/h_tpxo9.v1.tar.gz?dl=1
> Would you like this software to fetch and stage the TPXO file from the internet now?
>  [Y/n] 
> --------------------- >> end captured stdout << ----------------------
> 
> ----------------------------------------------------------------------
> Ran 63 tests in 4.025s
> 
> FAILED (errors=1)
> Error: Process completed with exit code 1.
> ```
> 
> is there a way to auto-accept the download?

We might not want to do that, because this is a large file. Is there a way to provision the GitHub workflow image with the data?

File: tests/test_configuration.py [hunk 1]
```diff
@@ -83,16 +83,6 @@ def test_configuration(self):
...
-        # init tidal forcing and setup requests
-        tidal_forcing = Tides()
-        tidal_forcing.use_constituent('M2')
-        tidal_forcing.use_constituent('N2')
-        tidal_forcing.use_constituent('S2')
-        tidal_forcing.use_constituent('K1')
-        tidal_forcing.use_constituent('O1')
-
-        mesh.add_forcing(tidal_forcing)
-
```
Comment (ghost) [inline tests/test_configuration.py]: I removed tidal forcing for now so it doesn't download TPXO, but in the future we could use another call of https://github.com/actions/cache just like we currently do for dependencies

**Comment by ghost** [comment]
testing `fort.15` with NWS / NRS set to 517

File: adcircpy/fort15.py [hunk 2]
```diff
@@ -582,15 +623,50 @@ def NSCREEN(self):
...
+        elif self.mesh is not None:
+            wind_forcing = self.mesh._surface_forcing['imetype']
+            if wind_forcing is not None:
                 # check for wave forcing here as well.
-                return self.wind_forcing.NWS
+                nws = int(wind_forcing.NWS % 100)
+        else:
+            nws = 0
+
+        return nws
```
Comment (ghost) [inline adcircpy/fort15.py:664]: NWS parameter extracted from tens place of input configuration coming from the wind forcing class

File: adcircpy/fort15.py [hunk 3]
```diff
@@ -582,15 +623,50 @@ def NSCREEN(self):
...
+                nrs = wave_forcing.NRS
+            elif wind_forcing is not None:
+                nrs = int(math.floor(wind_forcing.NWS / 100))
             else:
-                return 0
+                nrs = 0
+        else:
+            nrs = 0
+
+        return nrs
```
Comment (ghost) [inline adcircpy/fort15.py:691]: NRS parameter extracted from wave forcing class OR from hundredths place of wind forcing NWS configuration

File: adcircpy/forcing/waves/base.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+
+class WaveForcing(Forcing):
+    def __init__(self, nrs: int, interval_seconds: int):
+        self.NRS = nrs
+        self.RSTIMINC = interval_seconds
+
+    def write(self, directory: PathLike, overwrite: bool = False):
+        # TODO implement this
+        pass
```
Comment (ghost) [inline adcircpy/forcing/waves/base.py:13]: wave forcing class implementing `.NRS` and `.RSTIMINC`

File: adcircpy/forcing/base.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from abc import ABC, abstractmethod
+from os import PathLike
+
+
+class Forcing(ABC):
+    @abstractmethod
+    def write(self, directory: PathLike, overwrite: bool = False):
+        raise NotImplementedError
```
Comment (ghost) [inline adcircpy/forcing/base.py:8]: abstract forcing class

File: adcircpy/forcing/winds/base.py [hunk 6]
```diff
@@ -1,2 +1,13 @@
...
+
+
+class WindForcing(Forcing):
+    def __init__(self, nws: int, interval_seconds: int):
+        self.NWS = nws
+        self.WTIMINC = interval_seconds
+
+    def write(self, directory: PathLike, overwrite: bool = False):
+        # TODO implement this
+        pass
```
Comment (ghost) [inline adcircpy/forcing/winds/base.py:13]: wind forcing class implementing `.NWS` and `.WTIMINC`

File: adcircpy/fort15.py [hunk 7]
```diff
@@ -1,246 +1,283 @@
...
         # ----------------
-        f = f'{self.RUNDES}'.ljust(63) + ' ! RUNDES\n'
-        f += f'{self.RUNID}'.ljust(63) + ' ! RUNID\n'
-        f += f'{self.NFOVER:d}'.ljust(63) + ' ! NFOVER\n'
-        f += f'{self.NABOUT:d}'.ljust(63) + ' ! NABOUT\n'
-        f += f'{self.NSCREEN:d}'.ljust(63) + ' ! NSCREEN\n'
-        f += f'{self.IHOT:d}'.ljust(63) + ' ! IHOT\n'
-        f += f'{self.ICS:d}'.ljust(63) + ' ! ICS\n'
-        f += f'{self.IM:d}'.ljust(63) + ' ! IM\n'
+        f = []
```
Comment (ghost) [inline adcircpy/fort15.py:27]: use list and `.join()` instead of incremental string concatenation

File: adcircpy/fort15.py [hunk 8]
```diff
@@ -1,246 +1,283 @@
...
-            f += ' ! ESL - LATERAL EDDY VISCOSITY COEFFICIENT\n'
-        else:
-            f += ' ! smagorinsky coefficient\n'
-        f += f'{self.CORI:G}'.ljust(63) + ' ! CORI\n'
+            interval = f'{self.WTIMINC}'
+            description = 'WTIMINC - meteorological data time increment'
+            if self.NRS in [1, 3, 4, 5]:
+                interval += f' {self.RSTIMINC}'
+                description += ', RSTIMINC wave forcing increment'
+            f.append(f'{interval:<63} ! {description}')
```
Comment (ghost) [inline adcircpy/fort15.py:77]: also append `RSTIMINC` if applicable

File: adcircpy/fort15.py [hunk 9]
```diff
@@ -1,246 +1,283 @@
...
-        f += f'{self.NTIP:d}'.ljust(63) + ' ! NTIP\n'
-        f += f'{self.NWS:d}'.ljust(63) + ' ! NWS\n'
-        f += f'{self.NRAMP:d}'.ljust(63) + ' ! NRAMP\n'
-        f += f'{self.G:G}'.ljust(63) + ' ! gravitational acceleration\n'
-        f += f'{self.TAU0:G}'.ljust(63) + ' ! TAU0\n'
+        f.extend(f'{attribute:<63}' for attribute in attributes)
+        f.extend([
+            f'{self.NCOR:<63d} ! NCOR',
+            f'{self.NTIP:<63d} ! NTIP',
+            f'{int((self.NRS * 100) + self.NWS):<63d} ! NWS',
```
Comment (ghost) [inline adcircpy/fort15.py:55]: combine `NRS` and `NWS` into single parameter

File: adcircpy/server/driver_file.py [hunk 10]
```diff
@@ -149,12 +149,12 @@ def _run_hotstart_phase(self):
             f += 'ln -sf ../coldstart/fort.67\n'
 
         if self._driver.wind_forcing is not None:
-            if self._driver.wind_forcing.NWS in [19, 20]:
+            if self._driver.NWS in [17, 19, 20]:
```
Comment (ghost) [inline adcircpy/server/driver_file.py:152]: add `17` to supported NWS values

File: examples/example_4.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+    tidal_forcing = Tides()
+    tidal_forcing.use_all()
+
+    wind_forcing = WindForcing(17, 3600)
+    wave_forcing = WaveForcing(5, 3600)
+
+    mesh.add_forcing(tidal_forcing)
+    mesh.add_forcing(wind_forcing)
+    mesh.add_forcing(wave_forcing)
```
Comment (ghost) [inline examples/example_4.py:38]: add wave and wind forcing to driver in new example

File: tests/data/reference/test_configuration/fort.15.hotstart [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+2                                                               ! ICS
+511113                                                          ! IM
+2                                                               ! NOLIBF
+2                                                               ! NOLIFA
+1                                                               ! NOLICA
+1                                                               ! NOLICAT
+0                                                               ! NWP
+1                                                               ! NCOR
+1                                                               ! NTIP
+517                                                             ! NWS
```
Comment (ghost) [inline tests/data/reference/test_configuration/fort.15.hotstart:16]: test for combined `NWS` parameter

File: tests/data/reference/test_configuration/fort.15.hotstart [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+8                                                               ! NRAMP
+9.81                                                            ! gravitational acceleration
+0.005                                                           ! TAU0
+2.510781                                                        ! DTDP
+0                                                               ! STATIM
+0                                                               ! REFTIM
+3600 3600                                                       ! WTIMINC - meteorological data time increment, RSTIMINC wave forcing increment
```
Comment (ghost) [inline tests/data/reference/test_configuration/fort.15.hotstart:23]: test for appended `RSTIMINC`

File: adcircpy/driver.py [hunk 14]
```diff
@@ -20,17 +22,16 @@
...
             mesh: AdcircMesh,
             start_date: datetime,
             end_date: datetime,
             spinup_time: timedelta = None,
             netcdf: bool = True,
-            server_config: Union[int, SSHConfig, SlurmConfig] = None
+            server_config: Union[int, SSHConfig, SlurmConfig] = None,
     ):
-        self._mesh = mesh
+        super().__init__(mesh)
```
Comment (ghost) [inline adcircpy/driver.py:34]: initialize with new super constructor in `Fort15` class

File: adcircpy/driver.py [hunk 15]
```diff
@@ -475,9 +487,29 @@ def tidal_forcing(self):
 
     @property
     def wind_forcing(self):
-        return self.mesh._surface_forcing['imetype']
+        if self.mesh is not None and self.mesh._surface_forcing is not None:
+            self.__wind_forcing = self.mesh._surface_forcing['imetype']
+        return self.__wind_forcing
```
Comment (ghost) [inline adcircpy/driver.py:493]: guard against missing mesh / forcing

File: adcircpy/driver.py [hunk 16]
```diff
@@ -475,9 +487,29 @@ def tidal_forcing(self):
...
         # if isinstance(elevbc, WindForcing):
 
+    @wind_forcing.setter
+    def wind_forcing(self, wind_forcing: WindForcing):
+        if self.mesh is not None and self.mesh._surface_forcing is not None:
+            self.mesh._surface_forcing['imetype'] = wind_forcing
+        self.__wind_forcing = wind_forcing
```
Comment (ghost) [inline adcircpy/driver.py:500]: set wind forcing in dictionary to be `WindForcing` class

File: adcircpy/driver.py [hunk 17]
```diff
@@ -475,9 +487,29 @@ def tidal_forcing(self):
...
+
+    @property
+    def wave_forcing(self):
+        if self.mesh is not None and self.mesh._boundary_forcing is not None:
+            self.__wave_forcing = self.mesh._boundary_forcing['iwrtype']['obj']
+        return self.__wave_forcing
```
Comment (ghost) [inline adcircpy/driver.py:506]: guard against missing mesh / forcing

File: adcircpy/driver.py [hunk 18]
```diff
@@ -475,9 +487,29 @@ def tidal_forcing(self):
...
+
+    @wave_forcing.setter
+    def wave_forcing(self, wave_forcing: WaveForcing):
+        if self.mesh is not None and self.mesh._boundary_forcing is not None:
+            self.mesh._boundary_forcing['iwrtype']['obj'] = wave_forcing
+        self.__wave_forcing = wave_forcing
```
Comment (ghost) [inline adcircpy/driver.py:512]: set wave forcing in dictionary to be `WaveForcing` class

File: adcircpy/driver.py [hunk 19]
```diff
@@ -577,18 +609,13 @@ def concentration_surface_output(self):
     def output_collection(self):
         return self._output_collection
 
-    @property
-    def wave_forcing(self):
-        return self.mesh._boundary_forcing['iwrtype']["obj"]
```
Comment (ghost) [inline adcircpy/driver.py:582]: move wave forcing property closer to wind forcing property

File: tests/test_configuration.py [hunk 20]
```diff
@@ -69,6 +71,49 @@ def test_slurm_driver(self):
...
+
+        driver.write(output_directory, overwrite=True)
+
+        for reference_filename in reference_directory.iterdir():
+            generated_filename = output_directory / reference_filename.name
+            with open(generated_filename) as generated_file:
+                with open(reference_filename) as reference_file:
+                    generated_lines = generated_file.readlines()[1:]
+                    reference_lines = reference_file.readlines()[1:]
+                    assert generated_lines == reference_lines
```
Comment (ghost) [inline tests/test_configuration.py]: test generated configuration files against reference

**Pull request merged**